appendices/spirvenv.txt

```diff
@@ -491,6 +491,8 @@ endif::VK_VERSION_1_1[]
   * [[VUID-{refpage}-HitAttributeKHR-04703]]
     A variable with code:HitAttributeKHR storage class must: only be written
     to in an intersection shader
+  * A variable with code:HitAttributeKHR storage class must: only be read
+    from in an any-hit or closest hit shader
```

Comment:
Also intersection (it can be read after being written). 
Depending on exactly how 4703 is parsed, it could be added up there instead..

Reply:
well "read/written" are taken from the SPIR-V spec

> They can be written to only in IntersectionKHR execution model and read from only in AnyHitKHR and ClosestHitKHR execution models.

but in practice, it turned into `OpLoad`/`OpStore` in the `spirv-val` PR

I guess what is a use case to read the value from the `HitAttributeKHR` variable instead of the variable used to write to it?

Reply:
I don't believe we have anything in SPIR-V that prohibits from reading back output variables, and we definitely want to allow that at the HLL. In theory you could do a compiler transform that does what you suggested, but I doubt that glslang does that, and I think that is getting us off into the weeds here. 

Looking back at the GLSL we say: 

    4.3.X hitAttributeEXT Variables

    These are allowed only in intersection, any-hit, and closest-hit shaders.
    It is a compile-time error to use them in any other stage. They can be read
    in any-hit, and closest-hit shaders. They can be written to and read from
    only in intersection shaders (values are undefined before being written to).

In the SPIR-V we say: 

    HitAttributeKHR
    ...
    Variables declared with this storage class are allowed only in IntersectionKHR,
    AnyHitKHR and ClosestHitKHR execution models. They can be written to only
    in IntersectionKHR execution model and read from only in AnyHitKHR and
    ClosestHitKHR execution models.

So the intent is to clearly allow R/W in Intersection and R in AnyHit and ClosestHit.

Looking back up a few more lines - we have VUID 04701 which already restricts HitAttributeKHR to Intersection, AnyHit and ClosestHit stages.
So 04703 is trying to say that Intersection is the only stage that is allowed to write to HitAttributeKHR (maybe we need slightly different wording to clarify that).
And your new VU should be trying to say that the only stages allowed to read from HitAttributeKHR are Intersection, AnyHit, and Closest hit.

How about: 

  * [[VUID-{refpage}-HitAttributeKHR-04703]]
    A variable with code:HitAttributeKHR storage class must: not be written to in any stage which is not an intersection shader.
  * A variable with code:HitAttributeKHR storage class must: not be read from in any stage which is not an intersection, any-hit, or closest hit shader.

But then the newly proposed one seems pretty redundant with the existing 04701 VU, so I think my only suggestion here might be to try to reword the 04703 VU, so that it's clear that "writing is only permitted in IS" and NOT "is write-only in IS"

Reply:
> They can be written to only in IntersectionKHR execution model and read from only in AnyHitKHR and ClosestHitKHR execution models. 

ok... I slightly misread that and "you can ONLY read from those 2 stages" not it was "read only IN those stages"

|   | Any | Closest | Intersection |
|---|-----|---------|--------------|
| W |     |         | X            |
| R | X   | X       | X            |

I think both the `4703` and `4701` already cover this then already! Lets just remove this line and maybe reword the SPIR-V spec to just say

> They can be written to only in IntersectionKHR execution model.

as adding the "read from only" is not need as it would be implied from the above sentence stating it is allowed in all 3 stages